### PR TITLE
Avoid disk I/O in compiler

### PR DESCRIPTION
**Pull request by leofang**
A very experimental attempt that is partially related to #3144. This feature might be useful for:
1. When I/O is a concern (say, due to slow filesystems).
2. Long-running GPU jobs that can reuse the in-memory cached kernels many times. (Short jobs will have no advantage, as the kernels get recompiled every time a job is launched.)
3. Hosts that have no memory constraint.
4. Easing write contention among multiple processes.

Only NVRTC's output can be cached in memory. NVCC (and HIPCC?) will just write to disk and so it's meaningless I guess. 

I do see a bit of performance improvement with the in-memory cache, but it likely depends on how the test script is crafted. As long as there's no performance degradation (didn't observe any) we should be fine. 

(One thing may be worth noting: In Numba, the CUDA kernels are also effectively cached in memory. But there it's due to JIT.)

I would not push this PR further unless stakeholders are on board. Comments, suggestions, and criticisms are more than welcome! 🙂 

cc: @jakirkham 

TODO:
- [x] Tests. Most likely we can only run `test_raw.py` twice, with the new env var `CUPY_CACHE_IN_MEMORY` set and unset? Can't think of a better way to test this...

(Locally I clean up and then monitor the `kernel_cache` directory to ensure I don't generate any files. I also monitor the number of entries in the cache `cupy.cuda.compiler._in_memory_cubin_cache`.)
- [x] Avoid writing source files to temp directories? Haven't figured out how it can be avoided...

### CODE REVIEW

**Comment by leofang** [comment]
Items on the todo list are all crossed.

**Comment by leofang** [comment]
Closing this as per discussion starting in https://github.com/cupy/cupy/issues/3144#issuecomment-597464683.

**Comment by leofang** [comment]
Reopen to prepare for addressing https://github.com/cupy/cupy/issues/3144#issuecomment-643713161.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by kmaehashi** [comment]
Is this still WIP?

**Comment by leofang** [comment]
@kmaehashi This is done, but I didn't have time to do benchmarks. Should I mark it ready for review?

**Comment by kmaehashi** [comment]
> @kmaehashi This is done, but I didn't have time to do benchmarks. Would you like me to mark it ready for review?

Ok, we were just revisiting old PRs. Please continue working on benchmark when you have time 👍

**Comment by leofang** [comment]
I think it would be nice to have this capability included in v8 regardless of performance (see justification below), but I am not very happy about the current state of both the implementation and tests. @asi1024 @kmaehashi Any suggestions?

Quoting the motivation mentioned in the PR description above: 

> 1. When I/O is a concern (say, due to slow filesystems).

Today when testing CuPy on Ascent (Summit's test cluster) with @danielballan and others, we kept running into failures because its GPFS crashed several times due to stress tests running on other nodes. It would be great to be able to compile solely in memory if needed.

**Comment by leofang** [comment]
OK with #3749 I add a test to detect if any file is written to the temporary directory at the tear down phase. This makes me feel more comfortable. I could not find a way to mock the usage of `CUPY_CACHE_IN_MEMORY` as in #3749, though.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

File: tests/cupy_tests/core_tests/test_raw.py [hunk 1]
```diff
@@ -341,12 +342,20 @@ def use_temporary_cache_dir():
                 yield path
 
 
+_in_memory_cache = os.environ.get('CUPY_CACHE_IN_MEMORY')
+
+
 @testing.parameterize(*testing.product({
     'backend': ('nvrtc', 'nvcc'),
+    'in_memory': ('1', '0'),
 }))
```
Comment (asi1024) [inline tests/cupy_tests/core_tests/test_raw.py:351]: How about writing without `testing.product`?
```suggestion
@testing.parameterize(
    {'backend': 'nvrtc', 'in_memory': False},
    {'backend': 'nvrtc', 'in_memory': True},
    {'backend': 'nvcc', 'in_memory': False},
)
```

File: tests/cupy_tests/core_tests/test_raw.py [hunk 2]
```diff
@@ -341,12 +342,20 @@ def use_temporary_cache_dir():
...
 class TestRaw(unittest.TestCase):
 
     def setUp(self):
+        if self.backend == 'nvcc' and self.in_memory == '1':
+            self.skipTest('nvcc does not support in-memory cache')
+        os.environ['CUPY_CACHE_IN_MEMORY'] = self.in_memory
```
Comment (asi1024) [inline tests/cupy_tests/core_tests/test_raw.py:357]: Could you avoid mutating `os.environ`?

File: tests/cupy_tests/core_tests/test_raw.py (hunk 1 above)
Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py:351]: This is much nicer! Thanks.

File: tests/cupy_tests/core_tests/test_raw.py (hunk 2 above)
Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py:357]: I have a fresh thought this morning. Let me try if I can mock it with `functools.partial`. We need to mock the last argument in `_compile_with_cache_cuda(..., cache_in_memory)`.

Comment (asi1024) [reply tests/cupy_tests/core_tests/test_raw.py:358]: How about mocking `_get_bool_env_variable`?

File: tests/cupy_tests/core_tests/test_raw.py [hunk 3]
```diff
@@ -341,12 +342,21 @@ def use_temporary_cache_dir():
...
+@testing.parameterize(
+    {'backend': 'nvrtc', 'in_memory': False},
+    {'backend': 'nvrtc', 'in_memory': True},
+    {'backend': 'nvcc', 'in_memory': False},
+)
 class TestRaw(unittest.TestCase):
 
     def setUp(self):
+        if self.backend == 'nvcc' and self.in_memory == '1':
+            self.skipTest('nvcc does not support in-memory cache')
```
Comment (asi1024) [inline tests/cupy_tests/core_tests/test_raw.py:357]: This parameter combination is not listed now.

Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py:357]: Sorry I was still working on it, but got interrupted due to power outage (our area is very unstable lately). Will push once I get the chance.

File: tests/cupy_tests/core_tests/test_raw.py (hunk 2 above)
Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py:358]: This sounds better than what I have! Let me give it a try...

Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py:358]: How do I mock `_get_bool_env_variable` such that when `name='CUPY_CACHE_IN_MEMORY'` we can change its `default` argument, but a call to `_get_bool_env_variable('CUPY_CACHE_IN_MEMORY', False)` is still valid?

Comment (leofang) [reply tests/cupy_tests/core_tests/test_raw.py]: Take a look at 50b0430.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by asi1024** [comment]
Could you resolve conflicts?

**Comment by leofang** [comment]
Done.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM!

**Comment by leofang** [comment]
Thanks, @asi1024!

**Pull request merged**